core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -479,15 +478,20 @@ protected Channel nodeChannel(DiscoveryNode node, TransportRequestOptions option
 
     @Override
     public void disconnectFromNode(DiscoveryNode node) {
-        try (Releasable ignored = connectionLock.acquire(node.getId())) {
-            NodeChannels nodeChannels = connectedNodes.remove(node);
-            if (nodeChannels != null) {
-                try {
-                    logger.debug("disconnecting from [{}] due to explicit disconnect call", node);
-                    IOUtils.closeWhileHandlingException(nodeChannels);
-                } finally {
-                    logger.trace("disconnected from [{}] due to explicit disconnect call", node);
-                    transportServiceAdapter.raiseNodeDisconnected(node);
+        // this might be called multiple times, so do a lightweight check outside of the lock
+        NodeChannels nodeChannels = connectedNodes.get(node);
```

Comment:
we can not check out of lock. The problem is that we need to make sure that we have a stricit linearization of connection operation. If you call disconnect and it succeed you know you are disconnected and no ongoing work from before will end up reconnecting you (not how we only add the connected node to the map after a successful connection). I think the right solution here is to add timeouts to the connections done from the pings? maybe an easy way is to have a different connection timeout for "light" connections then we do for normal ones.